### ISSUE TITLE
Retrieving the state of a different reactor

### ISSUE THREAD

**Issue by revol-xut** [lf-lang/lingua-franca#767]
Retrieving the state of a different reactor
--------------------------------

## What is the problem ?

Hi, I am currently writing on some learning project of mine (see [alarm clock](https://github.com/revol-xut/lf-alarm-clock)) and I hit a major roadblock. 

This project consists of two reactos which are logically and structurally nicely separated. The first reactor (`Network`) handles all the network traffic in a separate thread and the second (`Clock`) manages events and a save file. But occasionally the `Network` reactor requires data from the `Clock` reactor (accessing the state) and this is surprising difficult. 

I sketched up the following example which is a modification of one of the examples I found in one of your papers. There the `Cockpit` want's to access the state of the `Door`.

```
target Cpp;

reactor Door {
    state open: bool;
    input get_state: void;
    output current_state: bool;
    
    reaction (get_state) -> current_state {=
        current_state.set(open);
    =}
}

reactor Cockpit {
    private preamble {=
        #include <thread>
    =}
    input get_door_state: void;
    input received_door_state: bool;
    output request_door_state: void;
    state state_copy: bool;
    state got_signal: bool;

    reaction (get_door_state) -> request_door_state {=
        request_door_state.set();
        // this needed to go into a separate thread to work
        // this is effectively creating a deadlock but the idea is clear
        while (not got_signal) {
            std::this_thread::sleep_for(5ms);
        }
        bool door_state = state_copy;
        got_signal = false;
        std::cout << "The Door is: " << door_state << std::endl;
    =}

    reaction (received_door_state) {=
        state_copy = *(received_door_state.get().get());
        got_signal = true;
    =}
}

main reactor Aircraft{
    door = new Door();
    cockpit = new Cockpit();
    cockpit.request_door_state -> door.get_state;
    door.current_state -> cockpit.received_door_state;

    reaction (startup) -> cockpit.get_door_state {=
        cockpit.get_door_state.set();
    =}
}
```
I find this solution quite dissatisfying because of it's massive overhead. There is always the option of hacky solution but I want to do it right.

Is there a simpler solution ? A feature I don't know something like `awaitable actions` or do I follow a completely wrong approach ?

**Comment by edwardalee** [comment]
The problem in general with retrieving the state of another reactor is that unless it is done carefully, it introduces nondeterminism. I agree that the mechanism in the example you cite is awkward, and indeed we have a proposed language extension on the table that effectively performs a procedure call on another reactor without introducing nondeterminism. This has been prototyped in the TS target, but not yet in the other target languages.

In some cases, the dependencies introduced by the remote procedure calls will lead to cyclic dependencies, in which case, it is not possible to allow such access without introducing nondeterminism. For such cases, it may be useful to introduce further language extensions where a programmer may explicitly state that nondeterminism is acceptable for the given application. This comes with a substantial cost in testability, however.

**Comment by revol-xut** [comment]
@edwardalee Thank you very much for your quick response. Is there some write up of this language feature ?

Ok than I will stick to this solution until a better one arises.

**Comment by lhstrh** [comment]
For the subroutines, have a look at chapter 4.5 of [my thesis](https://www2.eecs.berkeley.edu/Pubs/TechRpts/2020/EECS-2020-235.pdf).

**Comment by cmnrd** [comment]
I would like to point out that there is another solution. Instead of pulling the state from the Door (i.e. sending a request to get the state), the Door could also push its state, to the Cockpit and the Cockpit could internally keep track of it. This has the downside that we now have two places that manage state.

I wonder if stateful ports could be a solution here. I mean ports that are persistent and keep their last value until a new one is set. I believe this is how ports in C currently work for none-pointer types. Using such a port, the Door could make the internal state available in a deterministic way.

**Comment by revol-xut** [comment]
@cmnrd Yeah this also would be possible to leave out the request part and always send an update when the state of the door changes. This is quite clever, especially in error-prone systems where you want a response from the component that it performed the action you wanted.

But there are no persistent ports currently in cpp target. So I will just manually store it. But the combination of those two look appealing. If the user request the upcoming events I can just send the stored value inside the `Network` reactor and when there is any change in my 'Clock' I update the value in the 'Network' reactor accordingly.

For the people interested in how this approach changed the example from above:
```
target Cpp;

reactor Door {
    state open: bool(true);
    output current_state: bool;

    reaction (startup) -> current_state {=
        current_state.set(open);
    =}
}

reactor Cockpit {
    private preamble {=
        #include <thread>
    =}
    input updated_door_state: bool;
    input get_door_state: void;
    state state_copy: bool;

    reaction (get_door_state) {=
        bool door_state = state_copy;
        std::cout << "The Door is: " << door_state << std::endl;
    =}

    reaction (updated_door_state) {=
        state_copy = *(updated_door_state.get().get());
    =}
}

main reactor Aircraft{
    door = new Door();
    cockpit = new Cockpit();
    door.current_state -> cockpit.updated_door_state;

    reaction (startup) -> cockpit.get_door_state {=
        cockpit.get_door_state.set();
    =}
}
```
Thanks guys for constructive ideas. :D
